Registry functions return Poll to enable parallel fetching of index data

arlosi opened the pull request:
Adds `Poll` as a return type for several registry functions to enable parallel fetching of crate metadata with a future http-based registry.

Work is scheduled by calling the `query` and related functions, then waited on with `block_until_ready`.

This PR is based on the draft PR started by eh2406 here [#8985](https://github.com/rust-lang/cargo/pull/8985).

r? @Eh2406
cc @alexcrichton 
cc @jonhoo

arlosi commented:
I've updated the @jonhoo 's http registry prototype based on this change and the initial results look good. Not quite ready for PR yet though: https://github.com/arlosi/cargo/commit/dd5eb9ec99fa85078a3d216598d6fe3b7ceecc3c#diff-cb43b9ba868eedb392ea22664d59b4444dd7c4c4f5e641ab045aa861a38ffa4c

Eh2406 commented:
Thank you for picking this up!

> `/// Block until all outstanding Poll::Pending requests are Poll::Ready.`

I assumed we would have an API more like "Block until at least one outstanding request is Poll::Ready". But having thought about it the conditions for my API to be helpful are corner cases of corner cases. And if you're only making one request your API prevents needing a retry loop, `query` is guaranteed to return Ready after only one call to `block_until_ready`.

arlosi commented:
@alexcrichton yes, I'd like to get this completed! Any suggestions you have to move it forward would be awesome.

alexcrichton commented:
Ok so talking with @Eh2406 a bit yesterday one of the concerns is that while this infrastructure makes sense none of it's really being exercised yet (since everything is always "ready"). I think that we will actually want to change how remote sources (git/registry) work today to actually leverage this new functionality, however. Basically I think that the `query` method should never block, including the git/registry sources that we have today. Instead they would be updated to follow this methodology where `query` queries the internal caches and if something isn't found then it's "not ready" and records internally what needs to happen. Later on the "wait for everything" it'd actually do blocking work and network things like happens today. Does that make sense?

arlosi commented:
I'm finally coming back to this in the new year. @alexcrichton I'm hoping for some clarification on what you'd like here for the existing git model. Currently, the `query` method doesn't block in the git model. If work needs to be done (such as fetching the repo), then the query method will fail, and the caller would do an `update()` then re-try the query. See here for details:

https://github.com/rust-lang/cargo/blob/24783312150544d1bc9d1075c7f42ce7b2f35540/src/cargo/sources/registry/mod.rs#L695

All the blocking work currently happens in the `Source::update` method for git-based registries, and I'm not sure what we'd like to do with it. It could be changed to return `Poll`, then actually updating (git fetching) would be done in the `block_until_ready`. This would necessitate many changes in the callers -- and I'm not sure it really makes sense, since we'd just be adding loops calling `block_until_ready`.  Or we could re-design the callers of `update` as well, so that the meaning of `Source::update` is "ensure you have the latest data when running `query`". That would mean that no updating actually occurred until a query was started, and `block_until_ready` was called.

This change has the potential to become large, depending on how much of the design is changed and I want to make sure we get agreement on how we want it to look before doing the work.

alexcrichton commented:
Yes that's a good point, and also an indicator at how the current design is a bit sprawling... Currently there's basically an implicit "protocol" between the resovler and sources that the resolve will appropriately, for some definition of "appropriately" call `update` ahead of time before calling `query` if necessary. This means that for git sources the resolver determines that `update` needs to be called and invokes it.

Instead I think a better model would be to remove `update` entirely, folding the functionality into a `query` and `block_until_ready` split. This is indeed a large-ish change, and won't be trivial to make. I think, though, that this new model can be implemented today without changing the underlying implementations, just the interface to them largely.

If the same loop ends up being written in Cargo then that seems emminently solvable with a helper method or similar, I don't think we need to worry too much about the impact. The purpose of this change is to enable parallel downloads and updates for the resolver, and if everything else doesn't act in parallel that's ok we can update them later over time.

arlosi commented:
@alexcrichton I updated this PR to remove the `update` method as you suggested. 

If the registry has no local cache (hasn't been cloned yet, for instance), then queries will return `Poll::Pending`, and `block_until_ready` will do the cloning.

I also added an "invalidate_cache" method on the registry that makes subsequent queries return `Poll::Pending` even if a local clone exists. Calling `block_until_ready` then does the update and queries return results.

Eh2406 commented on src/cargo/sources/registry/remote.rs:
```diff
@@ -35,6 +36,8 @@ pub struct RemoteRegistry<'cfg> {
     head: Cell<Option<git2::Oid>>,
     current_sha: Cell<Option<InternedString>>,
+    needs_update: Cell<bool>,
+    updated: bool,
```
Generally `RemoteRegistry` is making my head spin. How do `needs_update` and `updated` interact? What code is new vs. what has just moved?

Do we have tests that the index is not fetched to much? I.E. At most once per invocation to `cargo` and none if the lockfile works.

I will look again when I am fresh and it may be clearer.

arlosi replied:
Most of the code just moved. It was previously in `update`, and had to move to outside the trait implementation into `update_index_worker`. 

There are some functional changes to `load`. Previously, it would error out if a crate was unavailable, then layers above would catch the error, `update`, and try again. Now `load` keeps track itself. It will only error out if an `update` has happened.

`needs_update` is associated with `invalidate_cache`. It means that someone as requested that this source must fetch the index for the current cargo session.

`updated` is to prevent fetching the index more than once.

Yes, there are tests covering those scenarios. I hit failures when I was implementing this :).

Eh2406 replied:
That makes sense now.

Eh2406 reviewed (commented):
Overall I think this is looking really good! Thank you for pushing this along.

arlosi commented on src/cargo/core/compiler/future_incompat.rs:
```diff
@@ -279,15 +280,35 @@ fn get_updates(ws: &Workspace<'_>, package_ids: &BTreeSet<PackageId>) -> Option<
+            }
+        });
+        for (_, source) in sources.iter_mut() {
+            source.block_until_ready().ok()?;
```
For a given source and query parameters, a `query` call that previously returned `Poll::Pending` must return `Poll::Ready` after calling `block_until_ready`. If the query parameters are different (such as querying for a package that hasn't been queried before), then `query` may return `Poll::Pending`, even after a `block_until_ready` call.

Calling `block_until_ready` if a source hasn't yet returned a `Poll::Pending` is an interesting question. If you call `invalidate_cache` + `block_until_ready`, the git-based sources will do a fetch. 

I agree we should be more precise here and only call `block_until_ready` for the sources remaining in `package_ids`

arlosi replied:
Ok, I've changed it to *should* return `Poll::Ready` after `block_until_ready` and updated the doc comments.

arlosi commented on src/cargo/sources/git/source.rs:
```diff
@@ -180,7 +133,67 @@ impl<'cfg> Source for GitSource<'cfg> {
+            src.query(dep, f)
+        } else {
+            Poll::Pending
+        }
```
I'd rather avoid starting the fetch in the background. It should be very quick for all the `Poll::Pending`s to return up, and have `block_until_ready` start doing work.

Eh2406 replied:
But if I understand it correctly then if you have multiple git dependencies they will be fetched serially when each of their `block_until_ready` is called. If it is triggered here the fetch can happen in parallel. Definitely something we can leave for later.

arlosi commented on src/cargo/sources/registry/remote.rs:
```diff
@@ -168,80 +221,76 @@ impl<'cfg> RegistryData for RemoteRegistry<'cfg> {
-        };
-        data(blob.content())
+        let mut not_found = false;
+        let result = || -> CargoResult<CargoResult<()>> {
```
Yes, what I really want is the unstable `try_blocks` feature. I'll move to helper function.

arlosi commented on src/cargo/core/registry.rs:
```diff
@@ -314,37 +321,55 @@ impl<'cfg> PackageRegistry<'cfg> {
-                    )
-                })?;
+
+                let summaries = match source.query_vec(dep)? {
```
This is nightly only as `try_trait_v2`, so we can't use it, right?

Eh2406 replied:
On stable you can not imple the trait in `try_trait_v2`. You can use `?`, even though it de-sugars to them.

arlosi replied:
Ok, so I am already using the try_trait_v2 here already. That's what's enabling `?` to cause `Poll::Ready(Err(...))` to return up. 

Maybe I'm missing something, but I don't see any way to use it further here.

Eh2406 replied:
Yes, sorry for the miss communication. Given how `?` works on `Poll` your code is correct. I was leaving a note that `?` did not do what I expected.

arlosi commented on src/cargo/sources/directory.rs:
```diff
@@ -17,6 +18,7 @@ pub struct DirectorySource<'cfg> {
     packages: HashMap<PackageId, (Package, Checksum)>,
     config: &'cfg Config,
+    updated: bool,
 }
```
I think that could cause an infinite updating loop if the directory source had no packages in it.

Eh2406 replied:
Good call.

arlosi commented on src/cargo/sources/git/source.rs:
```diff
@@ -19,6 +20,7 @@ pub struct GitSource<'cfg> {
     path_source: Option<PathSource<'cfg>>,
     ident: String,
     config: &'cfg Config,
+    updated: bool,
```
Done

alexcrichton commented:
I'm going to be [stepping down from the Cargo team](https://github.com/rust-lang/team/pull/712) so feel free to merge this without me, I unfortunately won't be able to get around to reviewing this.

bors commented:
:hourglass: Testing commit 0c07056bfeffee213e21cdf1e48875f59f834282 with merge a77ed9ba87bfeaf3c2758a5b291876c65f734c39...
<!-- homu: {"type":"BuildStarted","head_sha":"0c07056bfeffee213e21cdf1e48875f59f834282","merge_sha":"a77ed9ba87bfeaf3c2758a5b291876c65f734c39"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/5486551615?check_suite_focus=true)
Approved by: Eh2406
Pushing a77ed9ba87bfeaf3c2758a5b291876c65f734c39 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Eh2406","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/5486551615?check_suite_focus=true"},"merge_sha":"a77ed9ba87bfeaf3c2758a5b291876c65f734c39"} -->

ehuss commented:
This PR is causing errors when integrating upstream.  Can you take a look at it?  You can test in the rust-lang/rust repo by doing:

1. `git submodule update --init`
1. `git submodule update --remote src/tools/cargo`
2. `cargo update -p cargo`
3. Set `submodules=false` in `config.toml`
4. `./x.py test src/tools/cargo`

I am seeing the following errors:

```
failures:

---- install::install_semver_metadata stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --registry alternative --v
ersion 1.0.0+abc`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --registry alternative`
thread 'install::install_semver_metadata' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --registry alt
ernative`
error: stderr did not match:
1        -    Updating `alternative` index
2   1          Ignored package `foo v1.0.0+abc (registry `alternative`)` is already installed, use --force to override
3   2     warning: be sure to add [..]


other output:

', src/tools/cargo/tests/testsuite/install.rs:1947:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- install_upgrade::already_installed_updates_yank_status_on_upgrade stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.0`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.1`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.1`
thread 'install_upgrade::already_installed_updates_yank_status_on_upgrade' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.
1`
error: process exited with code 101 (expected 0)
--- stdout

--- stderr
error: cannot install package `foo`, it has been yanked from registry `crates-io`
', src/tools/cargo/tests/testsuite/install_upgrade.rs:830:10

---- install_upgrade::already_installed_exact_does_not_update stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.0`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --version=1.0.0`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo '--version=>=1.0.0'`
thread 'install_upgrade::already_installed_exact_does_not_update' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo '--version=>=1
.0.0'`
error: stderr did not match:
1        -    Updating `[..]` index
2   1          Ignored package `foo v1.0.0` is already installed[..]
3   2     warning: be sure to add [..]


other output:

', src/tools/cargo/tests/testsuite/install_upgrade.rs:779:10

---- install_upgrade::no_track stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --no-track foo`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --no-track foo`
thread 'install_upgrade::no_track' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --no-track foo`
error: stderr did not match:
1        -    Updating `[..]` index
2   1     error: binary `foo` already exists in destination `[..]/.cargo/bin/foo`
3   2     Add --force to overwrite


other output:

', src/tools/cargo/tests/testsuite/install_upgrade.rs:703:10

---- install_upgrade::registry_upgrade stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/cit/t1302/home/.cargo/bin/foo-1`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
thread 'install_upgrade::registry_upgrade' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
error: stderr did not match:
1        -    Updating `[..]` index
2   1          Ignored package `foo v1.0.0` is already installed[..]
3   2     warning: be sure to add [..]


other output:

', src/tools/cargo/tests/testsuite/install_upgrade.rs:155:10

---- install_upgrade::v1_already_installed_dirty stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
thread 'install_upgrade::v1_already_installed_dirty' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
error: expected to find:
   Compiling foo v1.0.1

did not find in output:
     Ignored package `foo v1.0.0` is already installed, use --force to override
warning: be sure to add `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/cit/t1308/home/.cargo/bin` to yo
ur PATH to be able to run the installed binaries
', src/tools/cargo/tests/testsuite/install_upgrade.rs:330:10

---- install_upgrade::upgrade_force stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --force`
thread 'install_upgrade::upgrade_force' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install foo --force`
error: stderr did not match:
1        -    Updating `[..]` index
2   1       Installing foo v1.0.0
    2    +    Updating `dummy-registry` index
3   3        Compiling foo v1.0.0
4   4         Finished release [optimized] target(s) in [..]
5   5        Replacing [..]/.cargo/bin/foo
6   6         Replaced package `foo v1.0.0` with `foo v1.0.0` (executable `foo`)
7   7     warning: be sure to add `[..]/.cargo/bin` to your PATH [..]


other output:

', src/tools/cargo/tests/testsuite/install_upgrade.rs:223:10

---- install_upgrade::multiple_report stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install one two three`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install one two three`
thread 'install_upgrade::multiple_report' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install one two three`
error: stderr did not match:
1        -    Updating `[..]` index
2   1          Ignored package `one v1.0.0` is already installed, use --force to override
3   2          Ignored package `two v1.0.0` is already installed, use --force to override
4        - Downloading crates ...
5        -  Downloaded three v1.0.1 (registry `[..]`)
6        -  Installing three v1.0.1
7        -   Compiling three v1.0.1
8        -    Finished release [optimized] target(s) in [..]
9        -   Replacing [..]/.cargo/bin/three
10       -   Replacing [..]/.cargo/bin/x
11       -   Replacing [..]/.cargo/bin/y
12       -    Replaced package `three v1.0.0` with `three v1.0.1` (executables `three`, `x`, `y`)
    3    +     Ignored package `three v1.0.0` is already installed, use --force to override
13  4          Summary Successfully installed one, two, three!
14  5     warning: be sure to add `[..]/.cargo/bin` to your PATH [..]


other output:

', src/tools/cargo/tests/testsuite/install_upgrade.rs:647:10

---- publish_lockfile::warn_package_with_yanked stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo generate-lockfile`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo package --no-verify`
thread 'publish_lockfile::warn_package_with_yanked' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo package --no-verify`
error: stderr did not match:
1   1        Packaging foo v0.0.1 ([..])
2        -    Updating `[..]` index
3        -warning: package `bar v0.1.0` in Cargo.lock is yanked in registry `crates-io`, consider updating to a version that is not yanked


other output:

', src/tools/cargo/tests/testsuite/publish_lockfile.rs:339:10

---- publish_lockfile::warn_install_with_yanked stdout ----
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --locked foo`
running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --force foo`
thread 'publish_lockfile::warn_install_with_yanked' panicked at '
test failed running `/home/eric/Proj/rust/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/cargo install --force foo`
error: stderr did not match:
1        -    Updating `[..]` index
2   1       Installing foo v0.1.0
    2    +    Updating `dummy-registry` index
3   3      Downloading crates ...
4   4       Downloaded bar v0.1.1 (registry `[..]`)
5   5        Compiling bar v0.1.1
6   6        Compiling foo v0.1.0
7   7         Finished release [optimized] target(s) in [..]
8   8        Replacing [..]/.cargo/bin/foo
9   9         Replaced package `foo v0.1.0` with `foo v0.1.0` (executable `foo`)
10  10    warning: be sure to add [..]


other output:

', src/tools/cargo/tests/testsuite/publish_lockfile.rs:404:10


failures:
    install::install_semver_metadata
    install_upgrade::already_installed_exact_does_not_update
    install_upgrade::already_installed_updates_yank_status_on_upgrade
    install_upgrade::multiple_report
    install_upgrade::no_track
    install_upgrade::registry_upgrade
    install_upgrade::upgrade_force
    install_upgrade::v1_already_installed_dirty
    publish_lockfile::warn_install_with_yanked
    publish_lockfile::warn_package_with_yanked
```

ehuss commented:
These errors are from running `cargo test --release`, it is not necessary to use the rust repo.

arlosi commented:
@ehuss looking into it now.

arlosi commented:
@ehuss Fix is available in #10482. I tested it locally in both debug and release.
